### PR TITLE
Add challenge for variadic generics

### PR DESCRIPTION
**Pull request by daya0576**
Add a new challenge for variadic generics for #60 

Reference:
1. https://peps.python.org/pep-0646/
2. https://www.youtube.com/watch?v=hAj3nGzeSiQ

### CODE REVIEW

**Comment by laike9m** [comment]
Thanks for contributing this new challenge!

I see a few things that can be improved, for example, the code is missing `## End of your code ##`, as described in:
https://github.com/laike9m/Python-Type-Challenges/blob/main/docs/Contribute.md#how-to-add-a-new-challange

Also, is it possible to remove the following code from `question.py`, as it seems to be part of the solution. Or we can put it in hints.md
```
T = TypeVar("T")
Ts = TypeVarTuple("Ts")

S1: TypeAlias = Tuple[float, int]
S2: TypeAlias = Tuple[*S1, str]
```

**Comment by daya0576** [comment]
Hi, thank you for your suggestions!

I have removed the code from `question.py`, but is it ok if the question code cannot run?

File: challenges/advanced-variadic-generics/solution.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        ...
+
+    def add_dimension(self, x: "T") -> "Array[*Ts, T]":
+        ...
+
+
+## End of your code ##
+a: Array[float, int] = Array()
+b: Array[float, int] = Array()
+print(a + b)
```
Comment (laike9m) [inline challenges/advanced-variadic-generics/solution.py:24]: You can use `assert_type`, see https://tinyurl.com/ypmacu4q

Ditto for other places

File: challenges/advanced-variadic-generics/question.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+Define a `Array` type that supports addition of arrays with the same shape.
+"""
+
+
+class Array:
+    def __add__(self, other):
+        ...
+
+    def add_dimension(self, x):
```
Comment (laike9m) [inline challenges/advanced-variadic-generics/question.py:12]: It's unclear from test code what this should return

File: challenges/advanced-variadic-generics/question.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+"""
+TODO:
+
+Define a `Array` type that supports addition of arrays with the same shape.
```
Comment (laike9m) [inline challenges/advanced-variadic-generics/question.py:4]: "Define an `Array` type that supports element-wise addition of arrays with identical dimensions and types."

Same for solution.

File: challenges/advanced-variadic-generics/question.py (hunk 2 above)
Comment (laike9m) [reply challenges/advanced-variadic-generics/question.py:12]: Actually we might just remove `add_dimension`, I don't think it's necessary. wdyt?

**Comment by laike9m** [comment]
> Hi, thank you for your suggestions!
> 
> I have removed the code from `question.py`, but is it ok if the question code cannot run?

This is fine. We don't actually run the code after all.

**Comment by laike9m** [comment]
Could you also add a `solution2.py` using Python 3.12's new generic syntax? Thanks.

File: challenges/advanced-variadic-generics/question.py (hunk 2 above)
Comment (daya0576) [reply challenges/advanced-variadic-generics/question.py]: Ok, `add_dimension` method is used to demonstrate type concatenation. Let's remove it to keep the code simple.

**Comment by daya0576** [comment]
> Could you also add a `solution2.py` using Python 3.12's new generic syntax? Thanks.

Hi, I have added the `solution2.py` using the latest generic syntax. 🎄🎄🎄

**Comment by laike9m** [comment]
Thanks!

**Pull request merged**